lib/mpl_toolkits/mplot3d/spherical.py

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/python
+
+# This module adds functionality to plot spherical polygons and is a result of
+# the discussion of issue #5294 (Handling Spherical Polygons) on github. The
+# following people contributed to this discussion:
+# Nikolai Nowaczyk, Tyler Reddy, Nicolas P. Rougier, Phillip Wolfram,
+# Thomas A Caswell
+
+import numpy as np
+from mpl_toolkits.mplot3d.art3d import Poly3DCollection
+
+
+class SphericalPolygon(object):
+    """
+     Plots a spherical polygon onto a sphere.
+    """
+
+    def __init__(self,
+                 vertices,
+                 tri=[[0, 1, 2]],
+                 center=np.array([0, 0, 0]),
+                 radius=1.,
+                 subdivisions=None):
+        """
+        Parameters
+        ----------
+        vertices: an array of points in 3D assumed to be on a sphere
+        tri: a list [[i,j,k], ...] of lists [i,j,k] of integers i,j,k, where
+            i, j, k are the indices of the vertices defining the face. (If left
+            empty, class will assume only 3 vertices in clockwise order.)
+        center: a point in 3D defining the center of the sphere
+        radius: a positive double representing the radius of the sphere
+        subdivisions: an integer defining how often the triangulation of
+                      the polygon is subdivided before it is plotted. (If left
+                      empty, the class will try to guess this value such that
+                      the result looks nice.)
+        """
+
+        self.vertices = vertices
+        self.center = center
+        self.radius = radius
+        self.subdivisions = subdivisions
+        self.tri = tri
+        self._refine_triangulation()
+
+    def _project_to_sphere(self, point):
+        """
+        Projects point onto the sphere defined by self.center and self.radius.
+
+        Parameters
+        ----------
+        point : array of floats of length ndim
+
+        returns: array of floats of length ndim
+        """
+
+        length = np.linalg.norm(point - self.center)
+        return (point - self.center) / length * self.radius + self.center
+
+    def _subdivide(self):
+        """
+         Subdivides the triangulation once by replacing every triangle
+         by four smaller triangles defined by the vertices of the original
+         triangle and the midpoints of its edges.
+        """
+
+        faces_subdivided = []
+
+        for face in self.tri:
+            # create three new vertices at the midpoints of each edge
+            a = self.vertices[face[0]]
+            b = self.vertices[face[1]]
+            c = self.vertices[face[2]]
+
+            # add vertices to list
+            k = len(self.vertices)
+            self.vertices = np.insert(self.vertices, k,
```

Comment:
This is slower than it could be. It appears that you are adding a fixed number of vertices (`3 * len(self.tri)`), so it's better to resize `self.vertices` once and assign the new values into the bigger array.